server/eventlog.go

```diff
@@ -182,3 +234,130 @@ func convertToPbEvents(hash crypto.Hash, events []attest.Event) []*pb.Event {
+					suffixAt = len(prefix)
+					break
+				}
+			}
+			if suffixAt == -1 {
+				return nil, fmt.Errorf("invalid prefix seen for PCR%d event: %s", index, rawData)
+			}
+			hasher.Write(rawData[suffixAt : len(rawData)-1])
+			if !bytes.Equal(event.Digest, hasher.Sum(nil)) {
+				// Older GRUBs measure "grub_cmd " with the null terminator.
+				// However, "grub_kernel_cmdline " measurements also ignore the null terminator.
+				hasher.Reset()
+				hasher.Write(rawData[suffixAt:])
+				if !bytes.Equal(event.Digest, hasher.Sum(nil)) {
+					return nil, fmt.Errorf("invalid digest seen for GRUB event log in event %d: %s", idx, hex.EncodeToString(event.Digest))
+				}
+			}
+			hasher.Reset()
+			commands = append(commands, string(rawData))
+		}
+	}
+	return &pb.GrubState{Files: files, Commands: commands}, nil
+}
+
+func getGrubKernelCmdlineSuffix(grubCmd []byte) int {
+	for _, prefix := range [][]byte{oldGrubKernelCmdlinePrefix, newGrubKernelCmdlinePrefix} {
+		if bytes.HasPrefix(grubCmd, prefix) {
+			return len(prefix)
+		}
+	}
+	return -1
+}
+
+// getDmVerityStateFromCmdline only supports the non-upstreamed ChromiumOS
+// dm-verity style arguments for now.
+// For example:
+// dm=1 vroot none ro 1,0 4077568 verity payload=<uuid> hashtree=<uuid> hashstart=4077568 alg=sha256 root_hexdigest=<digest> salt=<digest>
+// See:
+// https://source.chromium.org/chromiumos/chromiumos/codesearch/+/main:src/platform2/verity/README.md
+func getDmVerityStateFromCmdline(paramToVal map[string]string) (*pb.DmVerityState, error) {
+	if dmLine, ok := paramToVal["dm"]; ok {
+		dmArgs := parseArgs([]byte(dmLine))
+
+		hashAlg, ok := dmArgs["alg"]
+		if !ok {
+			return nil, fmt.Errorf("algorithm not specified in dm-verity configuration")
+		}
+		if hashAlg != "sha256" {
+			return nil, fmt.Errorf("invalid hash algorithm \"%v\" specified in dm-verity configuration: only sha256 is supported", hashAlg)
+		}
+
+		rootDigestStr, ok := dmArgs["root_hexdigest"]
+		if !ok {
+			return nil, fmt.Errorf("root digest not specified in dm-verity configuration")
+		}
+		rootDigestBytes, err := hex.DecodeString(rootDigestStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid hex specified for root digest in dm-verity configuration: %v", err)
+		}
+
+		saltStr, ok := dmArgs["salt"]
+		if !ok {
+			return nil, fmt.Errorf("salt not specified in dm-verity configuration")
+		}
+		saltBytes, err := hex.DecodeString(saltStr)
+		if err != nil {
+			return nil, fmt.Errorf("invalid hex specified for root digest in dm-verity configuration: %v", err)
+		}
+
+		return &pb.DmVerityState{HashAlg: tpmpb.HashAlgo_SHA256, RootDigest: rootDigestBytes, Salt: saltBytes}, nil
+	}
+	return nil, fmt.Errorf("supported dm-verity arg not found")
+}
+
+func getLinuxKernelStateFromGRUB(grub *pb.GrubState, parseVerity bool) (*pb.LinuxKernelState, error) {
+	var cmdline string
+	var err error
+	var verity *pb.DmVerityState
+	seen := false
+
+	for _, command := range grub.GetCommands() {
```

Comment:
What will happen if no kernel cmdline is found? Should throw an error?

Reply:
I don't know what issues there could be with GRUB potentially used with non-Linux systems or some Linux setup with an auto or empty command line. I also didn't want to add a "ParseKernelCmdline" option. That said, we can document that we only support certain common configurations (GRUB booting the kernel) and error here.